### ISSUE TITLE
Implement ssh port forwarding

### ISSUE THREAD

**Issue by pedromzadeh** [thebaselab/codeapp#764]
Would it be possible to implement SSH port forwarding? The _Blink_ app has this feature, so I think it should be achievable on iOS. 

For instance, `ssh -L 8888:localhost:8888 username@server_address`.

Would be pretty useful for spinning up and accessing remote Jupyter servers.

**Comment by thekeith** [comment]
@pedromzadeh I’d go a step further and add ssh config to enable forwarding of multiple ports and agent forwarding as well. I raised this in #928 before I saw your issue

**Comment by bummoblizard** [comment]
This is doable but it's going to involve a lot of work. We need to add the relevant features in our Objective-C libssh2 wrapper https://github.com/thebaselab/NMSSH to support this.

**Comment by bummoblizard** [comment]
https://libssh2.org/examples/tcpip-forward.html

**Comment by thekeith** [comment]
> This is doable but it's going to involve some work. We need to add the relevant features in our Objective-C libssh2 wrapper https://github.com/thebaselab/NMSSH to support this.

I’ll go dig up my Objective-C skills out of the closet…

**Comment by bummoblizard** [comment]
> I’ll go dig up my Objective-C skills…

@thekeith I am working on it.

**Comment by bummoblizard** [comment]
Proposed NMSSH API:
```swift
class NMSSHChannel {
  ...
  /* 
  Start listening on a local port and forward any TCP connection to the host, non-blocking
  queue: The queue where libssh2 calls are executed in (libssh2 and NMSSH are not thread-safe by design)
  */
  func bindLocalPortToRemoteHost(localListenIP: String, localPort: Int32, hostTargetIP: Int32, hostTargetPort: Int32, in queue: DispatchQueue) throws -> NMSSHSocket
}

class NMSSHSocket {
  // Underlying libssh2 socket
  private var sock: libssh2_socket_t 

  // To stop listening on a port
  func close()

  // TODO: Delegate / methods for status updates
}
```

`bindLocalPortToRemoteHost()`: 
- Initialise a POSIX socket
- Accept connections until the socket is closed from another thread. As `accept()` is a blocking call, it should be called in a new thread / `DispatchGroup`. 
- Once a connection is accepted, an auxiliary function is called in the provided DispatchQueue
- The function sets up a new libssh2 channel using `libssh2_channel_direct_tcpip_ex`
- Receive the available data from the socket and forward it to the remote host.
- Forward the response back to the local socket.
- Close the channel (A channel only serves a single TCP connection)

**Comment by bummoblizard** [comment]
![Simulator Screenshot - iPad (10th generation) - 2023-08-30 at 16 40 21](https://github.com/thebaselab/codeapp/assets/38398443/de072ec8-25b1-4cef-88a2-633095d328b9)

**Comment by thekeith** [comment]
Great to see progress here. How would it be configured?

One caveat is making sure that it doesn’t require a separate action to connect forwarded ports. In Termius for example you have to make a completely separate connection and initiate it for port forwarding which makes it a pain to set up, especially when you’re already in a session or want to forward multiple ports.

Ideally it would be great to allow users to do so via a UX which simply writes to a ~/.ssh/config file.

**Comment by bummoblizard** [comment]
> Great to see progress here. How would it be configured?
> 
> One caveat is making sure that it doesn’t require a separate action to connect forwarded ports. In Termius for example you have to make a completely separate connection and initiate it for port forwarding which makes it a pain to set up, especially when you’re already in a session or want to forward multiple ports.
> 
> Ideally it would be great to allow users to do so via a UX which simply writes to a ~/.ssh/config file.

I am thinking to support both GUI and config file. Let's support GUI first and deal with parsing .ssh/config in https://github.com/thebaselab/codeapp/issues/928.

**Comment by bummoblizard** [comment]
It will be available for testing in TestFlight after 1.5.0 is approved for public testing:
![Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2023-09-01 at 13 14 28](https://github.com/thebaselab/codeapp/assets/38398443/4c696d9d-518c-4232-a9b3-91cceca04182)
